### PR TITLE
TIP-706: Boolean/identifier sorters

### PR DESCRIPTION
**Pull request by SamirBoulil**
**Description (for Contributor and Core Developer)**

**ES Sorter Checklist:**
- [X] Add sorter integration tests for the field in the PQB
- [X] Add missing integration to the PimCatalogXIntegration tests for sort
- [X] Add the sorter + specs
- [ ] Update the reference documentation

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added Behats                      | -
| Added integration tests           | Ok
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst [hunk 1]
```diff
@@ -1308,7 +1308,7 @@ Equals (=)
 
     'filter' => [
         'term' => [
-            'values.description-text.<all_channels>.<all_locales>' => true
+            'values.a_yes_no-boolean.<all_channels>.<all_locales>' => true
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst:1341]: Oups :/

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php [hunk 2]
```diff
@@ -9,7 +9,7 @@
  * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
  */
-class CompletenessSorter extends BaseFieldSorter
+class CompletenessSorter extends IdentifierSorter
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php]: That is very strange to see that. No functional sense.
 I would prefer your Identifier sorter extend the BaseFieldSorter and just add what he need.

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/IdentifierSorter.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            );
+        }
+
+        $this->searchQueryBuilder = $searchQueryBuilder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addFieldSorter($field, $direction, $locale = null, $channel = null)
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/IdentifierSorter.php]: This looks to me identical to the BaseFieldSorter. Could you not just extend it and add `implements AttributeSorterInterface` + the two needed methods?

**Review by damien-carcel — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst [hunk 4]
```diff
@@ -516,6 +518,33 @@ Not In list
         ]
     ]
 
+Filtering
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst]: Sorting

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst [hunk 5]
```diff
@@ -516,6 +518,33 @@ Not In list
...
+~~~~~~~~~
+Operators
+.........
+
+Whenever one wants to sort on the field 'identifier' or an attribute of type 'pim_catalog_identifier'. The PQB will always sort on the field 'identifier'
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/search_specification.rst]: 'PQB' => The sort query will be performed on the field 'identifier'.

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/IdentifierSorter.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+namespace Pim\Bundle\CatalogBundle\Elasticsearch\Sorter;
+
+use Pim\Component\Catalog\Model\AttributeInterface;
+use Pim\Component\Catalog\Query\Sorter\AttributeSorterInterface;
+
+/**
+ * Identifier sorter for an Elasticsearch query.
+ *
+ * As the identifier product value is not indexed in ES. Whenever you want to sort on the attribute identifier 'sku'
+ * or on the field 'Identifier', the PQB will use this class which sorts on the field identifier.
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/IdentifierSorter.php]: Identifier => identifier

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/IdentifierSorter.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    public function __construct(array $supportedFields = [], array $supportedAttributes = [])
+    {
+        $this->supportedFields = $supportedFields;
+        $this->supportedAttributes = $supportedAttributes;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsField($field)
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/IdentifierSorter.php]: Remove this one

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Sorter/Attributes/Boolean/BooleanSorterIntegration.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                ]
+            ]);
+
+            $this->createProduct('no', [
+                'values' => [
+                    'a_yes_no' => [['data' => false, 'locale' => null, 'scope' => null]]
+                ]
+            ]);
+
+            $this->createProduct('empty_product', []);
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Sorter/Attributes/Boolean/BooleanSorterIntegration.php]: Add product with 'data' => null

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Sorter/Attributes/Boolean/LocalizableScopableSorterIntegration.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                    'a_localizable_scopable_yes_no' => [
+                        ['data' => true, 'locale' => 'en_US', 'scope' => 'ecommerce'],
+                        ['data' => true, 'locale' => 'en_US', 'scope' => 'tablet'],
+                        ['data' => true, 'locale' => 'fr_FR', 'scope' => 'ecommerce'],
+                        ['data' => false, 'locale' => 'fr_FR', 'scope' => 'tablet']
+                    ]
+                ]
+            ]);
+
+            $this->createProduct('product_two', [
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Sorter/Attributes/Boolean/LocalizableScopableSorterIntegration.php:44]: add case with data => null

**Review by jjanvier — APPROVED** [review]
mob review
GTM

**Comment by SamirBoulil** [comment]
Merged in with https://github.com/jjanvier/pim-community-dev/pull/129

**Pull request closed without merging**